### PR TITLE
Add environment variable expansion to keycloak.conf

### LINKED ISSUE #11283: Add environment variable expansion to keycloak.conf
From discussion #11126
Hi,

I tried to use the following expression in keycloak.conf:
log-level=${ROOT_LOGLEVEL:INFO}

I wanted to use the env var ROOT_LOGLEVEL to set the root log level to DEBUG, but this doesn't work, because for keycloak.conf only system properties are resolved. The same expression for another logger works in quarkus.properties.

I'm wondering whether it makes sense to support the same expressions in keycloak.conf as it is done for quarkus.properties?

_Originally posted by @danielFesenmeyer in https://github.com/keycloak/keycloak/discussions/11126_

---
So it makes sense and we should add Support, either by changing the `StringPropertyReplacer` to also take environment variables into account or by cutting out it's `replaceProperties` method to avoid side effects (its used in many places) and calling the new method in `PropertyMappingInterceptor`

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #11283

### CODE REVIEW

**Comment by pedroigor** [comment]
@DGuhr You should be fine by just running the tests I just added.

**Review by DGuhr — CHANGES_REQUESTED** [review]
Code LGTM, thanks 👍 

but could you please add sth to the configuration docs about the expression expansion? I think its a pretty neat feature and should be mentioned (like they do e.g. here in quarkus: https://quarkus.io/guides/config-reference#property-expressions )

tried locally, tests run, but what now also works is e.g.:

Nested expansion of env vars:
#log-level=${ROOT_LVL:${FALLBACK_LVL}} -> uses ROOT_LVL if defined and FALLBACK_LVL is not.

ENV VAR-Concatenation (I added IN / FO in these two and server started with log lvl info)
log-level=$IN_VAR$FO_VAR

Partial use of env vars in strings:
log-level=WA${RN_VAR} -> returned log lvl WARN. could e.g. come in handy for db-url, where the hostname changes per environment, but port etc stays the same.

If you want I can provide a separate PR for the docs later on, just tell me and I approve this one instead. :)

**Comment by pedroigor** [comment]
@DGuhr Please, let me know if the doc change works for you.

**Comment by DGuhr** [comment]
@pedroigor for me yes, but not for the build ;)

File: docs/guides/src/main/server/configuration.adoc [hunk 1]
```diff
@@ -162,6 +162,24 @@ To add the initial admin user using environment variables, set `KEYCLOAK_ADMIN`
...
 Once the first user with administrative rights exists, you can use the UI or the command line tool `kcadm.[sh|bat]` to create additional users.
 
+== Using placeholders
+
+You are able to use placeholders to resolve the configuration option value from environment variables.
+
+.Using a placeholder to resolve the host from an environment variable
+[source, bash]
+----
+db-url-host="${r"${MY_DB_HOST}"}"
```
Comment (DGuhr) [inline docs/guides/src/main/server/configuration.adoc:172]: these quotes are duplicated during merge, no? shouldn't it be `$r{"${MY_DB_HOST}"}`? Looking at the html output it now looks like this:
Screenshot 2022-04-14 at 12 59 27
which is wrongly double-quoted (or do I miss sth?)

File: docs/guides/src/main/server/configuration.adoc [hunk 2]
```diff
@@ -162,6 +162,24 @@ To add the initial admin user using environment variables, set `KEYCLOAK_ADMIN`
...
+----
+
+You are also able to default to a specific value if the environment variable can not be resolved:
+
+.Setting a default value
+[source, bash]
+----
+db-url-host="${r"${MY_DB_HOST:mydb}"}"
```
Comment (DGuhr) [inline docs/guides/src/main/server/configuration.adoc:180]: see comment above

**Review by DGuhr — APPROVED** [review]
LGTM :)

**Review by stianst — APPROVED** [review]

**Comment by danielFesenmeyer** [comment]
That's very helpful, thanks for implementing this enhancement :thumbsup:

**Pull request merged**